File: plugin/src/main/java/dev/httpmarco/polocloud/velocity/VelocityPlatform.java
/*
 * Copyright 2024 Mirco Lindenau | HttpMarco
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package dev.httpmarco.polocloud.velocity;

import com.velocitypowered.api.event.Subscribe;
import com.velocitypowered.api.event.proxy.ProxyInitializeEvent;
import com.velocitypowered.api.plugin.Plugin;
import com.velocitypowered.api.proxy.ProxyServer;
import com.velocitypowered.api.proxy.server.ServerInfo;
import dev.httpmarco.polocloud.RunningProxyPlatform;
import dev.httpmarco.polocloud.velocity.command.CloudCommand;
import dev.httpmarco.polocloud.velocity.listener.*;
import lombok.Getter;

import javax.inject.Inject;
import java.net.InetSocketAddress;

@Getter
@Plugin(id = "polocloud", name = "PoloCloud-Plugin", version = "1.0.0", authors = "HttpMarco")
public final class VelocityPlatform extends RunningProxyPlatform {

    private final ProxyServer server;

    @Inject
    public VelocityPlatform(ProxyServer server) {
        super(it -> server.registerServer(new ServerInfo(it.name(), new InetSocketAddress(it.hostname(), it.port()))),
                it -> server.getServer(it.name()).ifPresent(registeredServer -> server.unregisterServer(registeredServer.getServerInfo())));
        this.server = server;
    }

    @Subscribe
    public void onProxyInitialize(ProxyInitializeEvent event) {
        for (var registered : this.server.getAllServers()) {
            this.server.unregisterServer(registered.getServerInfo());
        }

        var eventManager = this.server.getEventManager();
        eventManager.register(this, new PlayerChooseInitialServerListener(this.server));
        eventManager.register(this, new PlayerDisconnectListener());
        eventManager.register(this, new ServerConnectedListener());
        eventManager.register(this, new PostLoginListener());
        eventManager.register(this, new ServerKickListener(this));

        var commandManager = this.server.getCommandManager();
        commandManager.register(commandManager.metaBuilder("cloud").build(), new CloudCommand());

        this.changeToOnline();
    }
}

File: addons/proxy/src/main/kotlin/dev/httpmarco/polocloud/addons/proxy/platform/velocity/VelocityPlatform.kt
package dev.httpmarco.polocloud.addons.proxy.platform.velocity

import com.google.inject.Inject
import com.velocitypowered.api.event.Subscribe
import com.velocitypowered.api.event.proxy.ProxyInitializeEvent
import com.velocitypowered.api.proxy.ProxyServer
import dev.httpmarco.polocloud.addons.proxy.ProxyAddon
import dev.httpmarco.polocloud.addons.proxy.platform.velocity.events.PlayerLoginEvent
import dev.httpmarco.polocloud.addons.proxy.platform.velocity.events.VelocityMotdUpdater
import org.slf4j.Logger
import java.io.File

class VelocityPlatform @Inject constructor(
    private val server: ProxyServer,
    private val logger: Logger
) {

    private lateinit var proxyAddon: ProxyAddon

    @Subscribe
    fun onProxyInitialization(event: ProxyInitializeEvent) {
        this.proxyAddon = ProxyAddon(File("plugins/polocloud"), true)
        val config = proxyAddon.config

        val commandManager = this.server.commandManager
        commandManager.register(
            commandManager.metaBuilder("polocloud").aliases(*this.proxyAddon.config.aliases().toTypedArray()).plugin(this).build(),
            VelocityCloudCommand(this.proxyAddon, server),
        )

        val eventManager = this.server.eventManager
        eventManager.register(this, VelocityMotdUpdater(this, server, config))
        eventManager.register(this, PlayerLoginEvent(this, server, config))

        VelocityTablistUpdater( this, server, config)
    }

    fun proxyAddon(): ProxyAddon {
        return this.proxyAddon
    }
}